### ISSUE TITLE
Add support for vector tiles

### ISSUE THREAD

**Issue by pauldendulk** [pauldendulk/Mapsui#152]
I did some research on vector tiles in [VectorTileToBitmapRenderer](https://github.com/OsmSharp/VectorTileToBitmapRenderer). To add basic vector tile support to Mapsui we need:

- Add a VectorTileLayer. 
- The VectorTileLayer should fetch tiles and store the downloaded bytes in a cache like a regular TileLayer.
- The VectorTileLayer.GetFeaturesInView should not return tiles but the extracted vector tiles as Mapsui Geometries. Look at the VectorTileToBitmapRenderer project to see how to extract them. Then translate the GeoJson classes to Mapsui geometries.
- Generic Styling should then automatically be possible through the Style on the layer.

**Comment by charlenni** [comment]
For me the interessting part of vector maps are the offline capability. Mapsforge has a very small file size for maps and there are many websites to download ready made maps for any location you could think about.

But if I understand it correct, Mapsforge and Mapzen deliver the same information: all features in the bounds of a tile by using normal Google naming convention. So Mapsforge is a special case of a VectorTileLayer. This part should work in MapsforgeSharp. Only the representation of the features should be changed.

The vector data is than rendered with a given style file (normally XML) which exists in different formats (see for example [here](https://github.com/mapsforge/mapsforge/blob/master/docs/Rendertheme.md). This style file isn't integrated into map or vector data.

Your suggestion, that we should cache the vector data makes the map slow. The time consuming part is the rendering. So why not rendering the vector data into tiles and cache the ready made tiles? All other things would than be the same for bitmap tiles and vector tiles. Only problem is the rendering of names. They should exist on an extra layer. Perhaps better two: one for street names and one for other names. This would be useful, if you want to rotate the map.

**Comment by charlenni** [comment]
Argh, while pressing the comment button I found the reason, why you want to draw the map each time new: blur tiles. Yes, with drawing tiles each time, zomming isn't a problem. Was that the reason?

**Comment by pauldendulk** [comment]
Yes, an important advantage is high quality when zooming in. Also, smaller size on disk when storing locally. Speed of rendering depends on the number of features in the vector tiles, could be faster, could be slower. I noticed some of the vector tile people assume it is faster in practice. We will have to find out. 

Offline vector tiles will be stored in MBTiles just like regular tiles. This is already possible, just no samples of this yet. Also there seems to be an issues with some format. There is an issue for this.

**Comment by charlenni** [comment]
Ok, than there shouldn't be a problem to replace MBTiles with Mapsforge file format. The only difference up to now seems the format of the features that Mapsforge returns. Good to know.

I see a problem with the rendering theme. It isn't trivial to do this. Perhaps Mapsforge provides an example, how to do this.

**Comment by pauldendulk** [comment]
I thinks you need to replace MBTiles, it just stored the raw bytes. The layer should interpret it as bitmap tile or vector tile.

**Comment by charlenni** [comment]
Mapsforge doesn't know bitmap tiles. It is a vector only file format. The rest is correct.

Getting the vector data will be easy, but render this data with a theme will be much work :)

**Comment by DiabloRusso** [comment]
Hi!
I did some research on the use of vector tiles in Mapsui! @pauldendulk , your project is magnificent for its flexibility and expandability!

- > Add a VectorTileLayer.
hgjgjgjkjg

**Comment by DiabloRusso** [comment]
Hi!
@pauldendulk, your project is magnificent for its flexibility and expandability! I did some research on the use of vector fragments in Mapsui:

- The first three issues: That's right, I went the same way:

```
    public class VectorTileLayer : TileLayer
    {
        public VectorTileLayer(ITileSource source) : base(source) {}

        public override IEnumerable<IFeature> GetFeaturesInView(BoundingBox box, double resolution)
        {
            var dictionary = new Dictionary<TileIndex, IFeature>();
            var levelId = BruTile.Utilities.GetNearestLevel(Schema.Resolutions, resolution);
            RenderGetStrategy.GetRecursive(dictionary, Schema, MemoryCache, box.ToExtent(), levelId);
            var sortedFeatures = dictionary.OrderByDescending(t => Schema.Resolutions[t.Key.Level].UnitsPerPixel);
            return sortedFeatures.Where(f => f.Value.Geometry is Raster)
                    .Select(f => PbfDataToMapuiFeatures(((IRaster) f.Value.Geometry).Data))
                    .SelectMany(f => f);
        }
    }
```
It may be more appropriate to inherit the VectorTileLayer from GroupTileLayer and convert each vector tile into Mapsui layers containing Mapsui features?

- Applying a style to a vector tile:
I use [Mapnik](https://github.com/mapnik/mapnik) with [mapbox vector tile](https://github.com/mapbox/mapnik-vector-tile) to produce vector tiles for my taste. To generate vector tiles, Mapnik needs a XML-style, that I convert from [CartoCSS map stylesheets for the Standard map layer on OpenStreetMap.org](https://github.com/gravitystorm/openstreetmap-carto). To read CartoCSS, I use the partial code from [here](https://github.com/MapSurferNET/MapSurfer.NET-CartoCSS). I'm in the beginning of the road.

The weak point is rendering, WPF is not able to draw so many geometries! To get rid of freezing, I replace MapRenderer with my class DirectXMapRenderer(I use only Windows and .Net 4.0) based on SlimDX!

Thanx! Good luck!

**Comment by charlenni** [comment]
Hi @DiabloRusso,

is it possible to find your project on GitHub?

**Comment by DiabloRusso** [comment]
Hi, @charlenni!
Sorry, my project is quite raw + I do not have possibility to publish it, but I can answer specific questions that are in my competence!
Thanx!

**Comment by DiabloRusso** [comment]
Hi, @charlenni !
My project is quite raw + I do not have the possibility to publish it, but I can answer specific questions that are in my competence!

**Comment by charlenni** [comment]
Ok, if I understand it correct, than the vector MBTiles contain vector data in PBF format for each tile. This is nearly the same as Mapsforge format: vector data is organized in tiles.

So it would be helpfull, to have an interface to get vector tile data for a given tile. This data is than rendered with an existing renderer with a style, given by a style file (like CartoCSS, MapCSS or Mapsforge) with JSON or XML file format and than cached for Mapsui.

@pauldendulk Are there any object (like features and styles) that could be used for this? Is Mapsui the right place or is this better done in BruTile?

**Comment by DiabloRusso** [comment]
@charlenni absolutely right! In PBF format and compress GZip(not necessarily, but often). I do not know anything about Mapsforge format, I follow to [mapbox format](https://www.mapbox.com/vector-tiles/specification/)

To read CartoCSS, the code can be taken [here](https://github.com/MapSurferNET/MapSurfer.NET-CartoCSS) and adapted to mapsui style!

**Comment by pauldendulk** [comment]
@charlenni I want to keep features or geometries out of BruTile. It would add way more complexity. I have to choose some geometry format. Styling features is also very complicated. My idea was to use BruTile to fetch the data as raw data stream/byte array. A VectorTileLayer in BruTile could transform that to Mapsui features.

**Comment by pauldendulk** [comment]
@DiabloRusso It looks like to do the transformation of PDF to Mapsui feature in the GetFeaturesInView with PbfDataToMapuiFeatures. You could do this in a IProvider and then assign that IProvider to the Mapsui Layer.DataSource. This layer will then create those feature in a background thread. Perhaps this will improve performance.

**Comment by pauldendulk** [comment]
@DiabloRusso  You implemented an entire MapRenderer? cool :)

My route to improve performance would be to use the hardware acceleration of SkiaSharp. This is already supported but I have not given it a try at all. It was not completely implemented the last time I checked, don't know the current status.

**Comment by pauldendulk** [comment]
Out of curiosity I looked into a proper solution for vector tiles. I experimented [here](https://github.com/OsmSharp/VectorTileToBitmapRenderer/commit/f779c4df801da5cfad3bacc7af9b3d1099a95341) with a VectorTileLayer. 

The conclusion is that the TileFetcher needs to be changed. Right now it is creating a feature and adding it to the memory cache. For vector tiles we dont want a feature with a bitmap but the raw bytes. Possible solutions:
1. Don't add the tile to the cache but return is to the TileLayer which could than add it as a feature with bitmap (as with the regular TileLayer) or transform it to a collection of feature with vector (as with the VectorTileLayer).
2. Pass the tile layer some construction method. This is not as clear as the previous one. TileLayer and TileFetcher become more entangled.
3. Always add a byte[] to the TileCache and turn it into what you want in the TileLayer.

A further complication for is that the VectorTileLayer needs a more advanced cache. In the Vector layer the cache is completely replaced (this is not an option for the VectorTileLayer) and the TileLayer adds the new tile to the cache uses the TileCache to clean up old data. The VectorTileLayer will need a clean up like TileLayer although it will add multiple feature that come with every tile. It will have to use the TileCache that constains Lists for the features that correspond with a tile.

**Comment by charlenni** [comment]
Somewhere you wrote, that the drawing is the most time consuming part. I assume, that getting the style for a given feature is much more time consuming than the drawing. You must search the correct style for each feature in the dictionary of styles you get from style file. So you need a good concept how to do this :)

**Comment by pauldendulk** [comment]
Perhaps, maybe we can measure that. There are a couple of things that are not super optimized right now. 
- The features need to be fetched from the cache.
- The bitmaps get created in the render loop (only once, but still).

What I would like to have eventually is that the renderer gets the features and style as one batch and just has to draw. Everything else should be prepared in a background thread.

**Comment by charlenni** [comment]
To check the process again: you want to create a bitmap tile from vector data and save this in a cache or you want to draw the tile each time directly to the canvas?

**Comment by pauldendulk** [comment]
For vector tiles I do not want to draw a bitmap. I want to draw the vectors directly to the canvas. At least this is the initial idea. We have to see how it performs. But then again we might have to optimize the vector renderer.

The process as I picture it:
- For a given extent we need to fetch the tiles from the source (a web service or an MbTiles db). 
- These tiles are returned to the specific layer (vector or bitmap) in the raw format as byte array. So, this could be a protobuf or image data.
- For VectorTileLayers these byte arrays need to be transformed to a list of features. 
- This list of features (that corresponds to one vector tile) need to be stored in a TileCache that uses a tile index to retrieve them. So, the cache needs to be of: ```MemoryCache<List<Feature>>```. Each list of Features belongs to a specific vector tiles. 
- All this needs to be done in the background thread.
- At GetFeaturesInView the features will have to be returned from the MemoryCache. The various lists need to be aggregated to a single set. For the renderer it will be no different than any other list of features it draws.

It will not be a lot of code to create this but it needs to be done without breaking the current functionality. I want to prevent duplication.

**Comment by DiabloRusso** [comment]
@pauldendulk I'm planning create VectorTileProvider how did you suggest(see [TileProvider](https://github.com/pauldendulk/Mapsui/blob/master/Mapsui/Providers/TileProvider.cs)):

- MemoryCache will contain Features instead of a byte[]

- Modify GetTileOnThread function like this:

```
private void GetTileOnThread(object parameter) 
        {
            var parameters = (object[])parameter;
            if (parameters.Length != 4)
                throw new ArgumentException("Four parameters expected");
            var tileProvider = (ITileProvider)parameters[0];
            var tileInfo = (TileInfo)parameters[1];
            var features = (MemoryCache<Features>)parameters[2];
            var autoResetEvent = (AutoResetEvent)parameters[3];

            try
            {
                features.Add(tileInfo.Index, CreateFeaturesFromVectorTileData(tileProvider.GetTile(tileInfo)));
            }
            catch (Exception ex)
            {
                Logger.Log(LogLevel.Error, ex.Message, ex);
                // todo: report back through callback
            }
            finally
            {
                _queue.Remove(tileInfo.Index);
                autoResetEvent.Set();
            }
        }
```

- For CreateFeaturesFromVectorTileData must provide styles for each feature

**Comment by DiabloRusso** [comment]
As for rendering...
Creating dynamic resources in a large number, such as geometry and brushes, to display on the device is very heavy operation and should not be done in a render loop. If brushes can be cached at the stage of reading styles, then geometry must be created on the fly at least once and placed in RenderedGeometry of Featute and except in the rendering loop I don't know where to do it. Need to find a way to create geometry asynchronously so that in the rendering loop only display them.

**Comment by pauldendulk** [comment]
Are you offering to implement this? If yes, I would like to help because it is difficult, in the heart of Mapsui. Maybe you can share your fork and branch.

About your code. Yes, this looks like what I have in mind. I assume though that you need a copy of the TileFetcher (the VectorTileFetcher?). I would like to minimize duplication but this could be a good starting point. 

About your second remark related to performance. In the WPF platform specific feature is cache in feature.RenderedGeometry. It is created in the render loop. But only once. It is not bad. But I prefer it os completely outside the render loop. The problem here is that creating the cached feature needs a platform specific renderer,  the question is where to do this . uhhh... this is basically what you describe in your remark. :)

**Comment by DiabloRusso** [comment]
I'm just starting to do it.
I'm planning to use SharDX for rendering, something I did on SlimDX, but this is freeze UI.
I will be happy to work with you!
Thanx!

**Comment by charlenni** [comment]
@pauldendulk To redraw the tile each time is only usefull, when you want to rotate/tilt a map. So I assume, that you work in this direction. Correct?

**Comment by pauldendulk** [comment]
@charlenni I am not sure I understand. There are two things. 1) generate the platform specific feature (Skia or Xaml) to be able to render it. 2) draw the feature to the canvas.

The feature has a style and a geometry that can be a vector or image.

Drawing the tile to the canvas is done on every pan, zoom and rotate.

**Comment by pauldendulk** [comment]
@DiabloRusso The platform specific renderer is quite isolated from the rest right now. I would be possible to add another render but this is something I do not want to support in the core project. It could be done at a later stage, or in a separate library.

**Comment by charlenni** [comment]
It takes some time to understand this :). Sorry.

A vector tile consists of vectors and tags for this vectors. MBTiles and Mapsforge (I thinks the most common offline map types) are organized in tiles, so that you get all vectors with tag inside one tile. Than you have a map style file, which maps the tags to styles. If you have read in the vector tile and the style file, you have features, which have styles and geometries. This features could now be drawn onto a bitmap canvas (tile render engine) or direct to screen canvas. Is this all correct?

**Comment by DiabloRusso** [comment]
@pauldendulk Making some implementation of vector tile layer, I came to the conclusion that it is necessary to convert layers of tile into Mapsui layers, because there is a need to [compose layers](https://tilemill-project.github.io/tilemill/docs/guides/comp-op/) with each other when rendering, Mapnik styles also supports compositing operations.
I see it this way: A vector layer is a set of something layers that contain tile features, some provider changes set of features in these layers depending on zoom level. Then, when rendering, the vector layers can be compose.

**Comment by TomQv** [comment]
I'm not sure, if mapbox vector tile is a good format for offline vector maps, because of the huge amount of tiles that woulf be needed to cover a wide range of zoom levels. I'm also still searching, didn't find something proper yet.

**Comment by pauldendulk** [comment]
I can not think of any alternative that would be much better. Any map representation would need some kind of vector. Things that could be improved on vanilla vector format are:
- Compression. MBTiles has compression in protobuf. Perhaps it can be optimized, perhaps there is some overhead because duplication between tiles but I see those are marginal optimizations.
- Simplification of geometries: This is possible by storing only the top levels of the tile pyramid. 

For any specific app you need to make choices in:
- How much detail you need.
- What area you want to cover.
- How much disk space you find acceptable.

**Comment by TomQv** [comment]
I have common mobile apps like OSMand or Orux in mind. Didn't find out yet, how their vector format is build, but the maps are quite compact and covering large regions up to highest level of detail.

**Comment by xivk** [comment]
This is perfectly doable with vector tiles, you just have to include all details in let's say zoom level 14 and use the zoom level 14 tiles to render the map at 15 and beyond...

**Comment by xivk** [comment]
So like paul already said :)

**Comment by TomQv** [comment]
> you just have to include all details in let's say zoom level 14 and use the zoom level 14 tiles to render the map at 15 and beyond...

is this the way, how they are doing it?

**Comment by charlenni** [comment]
You should look at the Mapsforge format. This is one of the smallest open documented format. See https://github.com/mapsforge/mapsforge/blob/master/docs/Specification-Binary-Map-File.md

**Comment by pauldendulk** [comment]
ah, it would definitely be interesting to learn what osmand is using. Since it is open source it might be easy to adopt their format. Perhaps their data files could be reused. 

But vector tiles will also be on the menu. A simple version will not be hard to implement.

**Comment by pauldendulk** [comment]
@charlenni The mapsforge format is also a tile format? An alternative to mapbox's?

**Comment by charlenni** [comment]
@pauldendulk yes, it is a tile oriented vector format. I converted the reader from Java to C# to use the Mapsforge files. So, if you could read Mapbox vector files and create map images from it, you should could use Mapsforge too.

**Comment by TomQv** [comment]
> I converted the reader from Java to C# to use the Mapsforge files. 
That's interesting, will have a look into this

**Comment by pauldendulk** [comment]
@charlenni  Oh, that could be really useful, also for other projects. Perhaps you could publish this as a stand alone open source project and publish it to nuget.

For mapbox tiles now use this project https://github.com/bertt/mapbox-vector-tile-cs by https://github.com/bertt/. Your project could serve a similar role.

**Comment by charlenni** [comment]
That said, you want to have GeoJSON as format for vector data like in the above mentioned code. Is this correct?

**Comment by pauldendulk** [comment]
No, but the mapbox-vector-tile library parses the vector tiles to the GeoJson classes of the [GeoJson.NET](https://www.nuget.org/packages/GeoJSON.Net) package. So this is no requirement.

**Comment by charlenni** [comment]
It would be helpfull, if you could say, how the interface for such vector tile provider should look like. This provider should get a TileInfo and return a List<Feature>? Something like 
````
List<Feature> GetTile(TileInfo tile)
````
So it should implement ITileProvider and ITileSchema from BruTile? And ITileSource too? And if we do it this way, is it good, because it is the same as bitmap sources? Should this than all go to BruTile? Or do you want to create a new project like VruTile, which contains all providers for vector tiles (MBTiles and Mapsforge, online and offline)? And should other things like styling go there too?

Mapsforge returns POIs and ways with Tags. If I understand it correct, than this POIs and Ways should converted to Features with Geometries (POIs to Points and Ways to Polygons). This Features contain Attributes, which are filled from Tags as key-value-pairs. Correct up to now?

But Mapsforge Ways have a Layer interger too. It is the layer tag of OSM. Where should this go? To the attributes?

**Comment by charlenni** [comment]
I started a new project like you suggested. You could find it [here](https://github.com/charlenni/Mapsui.VectorTiles).

I called it Mapsui.VectorTiles. At the end it should consist of 3 major parts:
1. A VectorTileProvider, which extracts the vector data from a stream. 
2. A VectorTileStyler, which delivers the style for a given Tag combination
3. A VectorTileSource, which creates a list of styled Features for Mapsui.

Up to now, only the VectorTileProvider works. It could handle streams from Mapsforge (transfer from original Java code) and Mapbox (project from [here](https://github.com/bertt/mapbox-vector-tile-cs). Both providers deliver a list of VectorTileLayers, which have one or more VectorTileFeatures. This VectorTileFeatures have one or more VectorTileGeometries and the Tags for this feature.

It was a little bit difficult to use Mapsui classes without Mapsui ;). So I added the Mapsui classes (Mapsui.Geometries) to the project. Perhaps you have a better idea.

I added an example for Mapbox MBTiles vector format files (SQLite database) with zipped ProtoBuf entries. Interessted could find it [here](https://github.com/charlenni/Mapsui.VectorTiles/blob/master/Mapsui.VectorTiles.Mapbox.Tests/MapboxVectorTileProviderTests.cs).

There is much work left, even for VectorTileProvider (properties, coordinat systems and so on). But it is a starter to discuss, if this is the right way.

**Comment by pauldendulk** [comment]
hi @charlenni, thanks a lot. I am busy with another project right now and cannot look at your project before next week. Sorry about that.

**Comment by charlenni** [comment]
No problem, today is the first day after holidays. So I didn't wait for your answer up to now ;)

**Comment by pauldendulk** [comment]
I finally started looking at your project. I will still have very little time during the summer, but it's a start. I will add issues to the project.

**Comment by charlenni** [comment]
That not correct. Mapsui has 2 different renderers: Xaml and Skia. There were 5 other renderers in the history.

It could be, that Skia isn‘t the best choice, but it is easy to handle. That‘s the reason why you choosed it. To create a complete new one is many work.

**Comment by AliFlux** [comment]
Yeah WPF rendering is an option. But it performed even worse than Skia when I benchmarked it. Skia sure is easy to handle, and that's why I worked on it. But for on-the-fly rendering, the graphics backend needs to be switched.

**Comment by charlenni** [comment]
Yes, Wpf is really slower than Skia, but Skia runs not only on Windows, but also on mobile platforms. I think, you haven't this in mind while creating your implementation of vector tiles.

I changed some code and now its faster. My problem in the moment isn't the render engine, but calculation of style. Because it changes with zoom, you have to calculate it each drawing cycle new. It takes half of the time of the render process. And a third up to the half of this style finding is the filter check. If you zoom in in my sample to see Europe, you have to style over 60 000 features. 

Another thing is overzoom. But I saw, that you solved this. So I will look into your code, to see, how this works.

**Comment by AliFlux** [comment]
The first prototype only works for Windows desktop. But UWP is planned for sure.

The overzoom was tricky though. The approach I use is that the map is (practically) scaled up after the max zoom (14). But that bought in a ton of performance related issues. So I implemented various clipping techniques for polygons, lines, and text to cut the rendering time and now it works seamlessly.

**Comment by AliFlux** [comment]
I did some profiling of my code. Here is the result:

![bench](https://user-images.githubusercontent.com/8333794/38780095-f8b89e9a-40ea-11e8-9e39-c834a2f2b5a4.PNG)

30% of the time is spent on loading the pbf from mbtiles and parsing it. 16% is spent on drawing. 15% is spent on parsing style and applying the filters.

I think that the new tile loading and filter applying, and parsing style is only done at integer zoom level. And this can be deferred in another background thread, so that the map views the vector scaled up version of the current tile, until the new tile is loaded.

But rendering functions such as drawing are done at every single frame, pan and zoom events. That's why they have to be heavily optimized, so that they keep showing the previous vector data in scaled up form, until the new tile is fetched. Google Maps and Mapbox are using the same technique apparently.

Another thing I noticed was that Mapbox is rendering tiles at 512x512, so that the number of tiles requested on panning the map are minimized. Is there a way to use 512px tiles in Mapsui?

**Comment by pauldendulk** [comment]
Wow, great work. I created a new issue for this:
https://github.com/pauldendulk/Mapsui/issues/378

**Comment by pauldendulk** [comment]
@AliFlux Your library generates a bitmap from the vector tile. A perfect solution would be to draw vectors directly to the canvas but I think your current solution is already very valuable and has some advantages as wel (it can be easily reused by other libraries this way). In Mapsui the fetching is already done on a background thread. And the vector-to-bitmap rendering in your library is also done on a background thread when used in Mapsui. As your performance measurements show the rendering is only a small part of the total time to generate the bitmap so I think other optimizations are not very important (although faster is always better). By looking at the code I would say it is not very hard to turn your library into a net standard component so that it can be used on all platforms.

**Comment by AliFlux** [comment]
You're right. I designed this library as an independent tile provider. So that it can be plugged into any map control needed, or even deployed as a map server for vector-to-raster tile serving (like tileserver-gl).

However I would love if we could somehow use this component for directly rendering on a map control. It would require deep integration with Mapsui and a possible architecture level rewrite, but I'm willing to help.

**Comment by charlenni** [comment]
I have a version of [VectorTiles](https://github.com/charlenni/Mapsui.VectorTiles) integrated in Mapsui with direct rendering. All features from Mapbox data (or better OpenMapTiles data) are loaded, converted to Mapsui features and than styled on the fly with a Mapbox GL style file (here with osm-liberty.json). There are changes to Mapsui, that aren't in the master. So I copied VectorTiles to a Mapsui [branch](https://github.com/charlenni/Mapsui/tree/VectorTiles) to test and debug. But normally you could use VectorTiles complete independently. It's a PCL library. It uses BruTile for fetching tiles (from MBTiles file or from HTTP source) and than extracts the Pbf data with Bert Temmes library. It doesn't use Skia. All drawings are done by Mapsui.

The main function MapboxGLStyler takes a Mapbox GL style file and extracts all layers (raster and vector) and creates for each a Mapsui layer with attached VectorTileProvider. Each of this layers gets the style layers contained in the Mapbox GL style file. With this, all features are styled on-the-fly regarding the zoom level. All symbols are exported to a special SymbolProvider (could be from one vector layer or more) and could than be rendered as any other layer. That's where I'm in the moment. Idea is, that this provider has a worker, that, if something changes, creates a new list of symbols, that should be drawn, and than calls the DataChanged event. I don't know, if this will work, but hope :)

There is much work to do and a long way to go. Perhaps we decide at the end, that any other drawing library is better. But I think in the moment Skia works. 

As you can see, VectorTiles contains even more formats. If the Mapbox GL part works, it should be possible to create a Mapsforge provider too. Files could be readed, but theme styler must be elaborated.

And up to than: any help is appreciated.

**Comment by jefffhaynes** [comment]
I've posted a basic rastering engine built on Win2D in UWP.  I was originally using it with the UWP MapControl, but it has a lot of limitations and since moving it to Mapsui, it performs much better.  Needs lots of work but it's a start if anyone is interested...

I also wrote a style processor to go with it so it will use OSM style definitions.  Note that an osm mbtiles db needs to be in AppData/Local/Packages/<packageid>/LocalState for this to work.

Path labeling is supported but definitely needs work.  One collision and I throw in the towel...

https://github.com/jefffhaynes/LocalMap

**Comment by charlenni** [comment]
Hi Jeff,

great work.

While reading the code, I have some questions:
- How many tiles could you raster in one second?
- Did you made a speed test between MS Drawing and Skia as render engine?
- In [TileRasterer](https://github.com/jefffhaynes/LocalMap/blob/9556b5c86ef7e14a77a5aa7959328f966fbf40e8/LocalMap/TileRasterer.cs#L84) you use only the first style for each feature. Is this correct? Because if I'm right, each feature could have one or more styles and features are drawn in order of their style apperence. This is the way, e.g. borders of streets are drawn.
- I didn't see any attempts to draw symbols. Is this correct?

Btw. it would be nice to see a rendered tile from your code. Is it possible to see the following [tile](https://b.tile.openstreetmap.org/17/68234/47797.png)? Thank you for your efforts.

**Comment by jefffhaynes** [comment]
I haven’t done any benchmarking but I can work on that. All your assumptions are correct. I think part of the problem is I don’t fully understand the semantics of drawing tiles so I’m certain there are multiple issues. The missing street boarders is a great point. I’m almost certain I am missing those and I wondered why it looked funny.

Let me work on it today and drawing that tile.

Thanks




________________________________

**Comment by charlenni** [comment]
Your problem is, that you draw each feature only with the first style you find. You should draw them in order if the style. The correct way is 

foreach Style in StyleFile
  foreach Feature in FeatureList
    if Style is valid for Feature
      draw Feature with Style
    end
  end
end

And don‘t do it the other way.

What you will see is, that it is much slower than your solution, because you have to check each Feature for each Style.

**Comment by jefffhaynes** [comment]
Got it. Let me try that and see what happens. It’s very possible the better way to do this would be to simply port Skia to Win2D.  I just didn’t know it existed until yesterday...


________________________________

**Comment by jefffhaynes** [comment]
Unfortunately the non-commercial open map tiles db only goes down to zoom 14 I think so I can't render the tile you linked.  Is this where I should be overzooming?

**Comment by charlenni** [comment]
Than try to make https://c.tile.openstreetmap.org/14/8529/5974.png.

And perhaps we could discuss this further in your issues. Makes the thread really long ;)

**Comment by jefffhaynes** [comment]
Absolutely!  There's clearly some work to be done :)

![image](https://user-images.githubusercontent.com/9734727/59159909-3464d500-8a9e-11e9-82cb-143e844bf580.png)

**Comment by AliFlux** [comment]
Hi. Great work!

@charlenni is right, the layers have to be drawn in the order specified by the style. I made the same mistake when I was making my rasterizer, and ended up rewriting a big portion of the code when I discovered this.

I've been studying mapping engines for quite some time. And figured out the way a true vector map could be made:

It has to be a 3D map. Every coordinate in X,Y,Z in 3d world could correspond to a latitude,longitude,altitude. Each tile has to be drawn on a 3D plane, that means each line and text has to be a tessellated into a 3D mesh. Once this capability is achieved, it would be pretty easy to implement features such as pitch, elevation, overzooming, collision, and even 3D buildings overlay.

**Comment by jefffhaynes** [comment]
Thanks!  Out of curiosity, why would it need to be a 3D map?  Couldn't you accomplish a similar thing in 2 or 2.5D?

**Comment by AliFlux** [comment]
As mentioned, it's going to be pretty easy implementing certain features such as pitch, elevation, overzooming, collision, and even 3D buildings overlay. Plus it would pave way to implement a full 3D map with elevation tiles as an extension of a vector map.

**Comment by charlenni** [comment]
@AliFlux Yes, a 3D map would be nice, but in the most cases a 2D map is enough. How often do you use the 3D view of a map? And there are some disadvantages. Beside the greater effort to get this working, you need bigger databases because of the additional altitude informations.

So from my side a 2D offline vector map would be a fine solution.

**Comment by AliFlux** [comment]
It's a cost-benefit tradeoff. I started making 2D map because it felt simpler, when I continued the development I found out that making it all on a 3D world would be much simpler. Right now I'm working on a high performance vector map for an enterprise client, and I'm so thankful that I started off on a 3D engine using SharpDX.

**Comment by jefffhaynes** [comment]
Many 2D platforms are effectively 3D platforms these days in terms of hardware acceleration, etc. I’m not sure there’s as big of a difference between the two as you might imagine.


________________________________

**Comment by AliFlux** [comment]
For pure C# there is. With 2D there are options like skiasharp, Win2D, WritableBitmapEx. With 3D it's basically pure WPF, SharpDX, Helixtoolkit. These engines are quite different from each other. In case of a high level engine like Unity, Unreal, etc. 2D drawing is practically tessellation on top of a 3D plane.

But then again, it's a cost-benefit trade off. You might pull of a decent vector map in pure 2D canvas.

**Comment by jefffhaynes** [comment]
But both WPF and UWP support 2/2.5/3D natively, no?  I’m not talking Win2D, etc. which are purely for 2D drawing.


________________________________

**Comment by AliFlux** [comment]
They do, but they have pretty bad performance. I use SharpDx/helixtoolkit because its much faster.
